File: Flyweight/Ruby/PlayerFactory.rb
require_relative "Terrorist"
require_relative "CounterTerrorist"

class PlayerFactory
    def initialize
        @players = {}
    end

    def createPlayer(type)
        if @players[type]
            return @players[type]
        else
            player = nil

            case type
                when "Terrorist"
                    player = Terrorist.new
                    @players[type] = player
                when "CounterTerrorist"
                    player = CounterTerrorist.new
                    @players[type] = player
            end

            return player
        end
    end
end

File: Flyweight/Javascript/PlayerFactory.js
import CounterTerrorist from "./CounterTerrorist.js";
import Terrorist from "./Terrorist.js";

class PlayerFactory {
  constructor() {
    this.players = {};
  }

  createPlayer(type) {
    if (this.players[type]) {
      return this.players[type];
    } else {
      let player = null;

      switch (type) {
        case "Terrorist":
          player = new Terrorist();
          this.players[type] = player;
          break;

        case "CounterTerrorist":
          player = new CounterTerrorist();
          this.players[type] = player;
          break;
      }

      return player;
    }
  }
}

export default PlayerFactory;